### DISCUSSION TITLE
GCC cross-compilation from M1 to Intel for macOS

### DISCUSSION THREAD

**Discussion by kornelski** [Homebrew/discussions#2499] [Everyday usage] [answered]
On x86-64 I've used to use Homebrew's `gcc-11` to build x86-64 executables with OpenMP support. Now I've switched to an M1 Mac, but I still need to build x86-64 C executables with OpenMP support. How can I do that? Is it even possible?

Homebrew docs say for cross-compilation there's `x86_64-elf-gcc` formula, but that's a *weird* compiler. It's totally not a cross-compiler equivalent of brew's normal `gcc`. It doesn't seem to support _anything_. Not just OpenMP is broken:

```
x86_64-elf-gcc -fopenmp
x86_64-elf-gcc: error: unrecognized command-line option '-pthread'
```

but I can't even compile a "hello world":

```
/opt/homebrew/opt/x86_64-elf-binutils/bin/x86_64-elf-ld: cannot find crt0.o: No such file or directory
/opt/homebrew/opt/x86_64-elf-binutils/bin/x86_64-elf-ld: cannot find -lc
```

and the "elf" part of the name makes me suspect it's not even for macOS mach-O executables. It doesn't support `-arch` flag either.

* Is the `x86_64-elf-gcc` usable for building x86-64 macOS executables? 
* Do I need to install extra libs/headers/SDK/some wrapper for it to behave like a normal `gcc` compiler?
* If it's not suitable for cross-compiling to Intel macOS, what's the recommended way? For example, should I install another copy of Homewbrew for Intel, running under Rosetta?

**Comment by carlocab** [comment, marked as answer]
`x86_64-elf-gcc` compiles [ELF](https://en.wikipedia.org/wiki/Executable_and_Linkable_Format) executables. macOS uses [Mach-O](https://en.wikipedia.org/wiki/Mach-O), not ELF.

If you need to compile for Intel on M1, use Clang. If you need to use GCC specifically, you _might_ be able to build your own cross-compiler, but that's not something Homebrew provides.

Using a Rosetta installation of Homebrew to install an Intel GCC will probably work too.
